include/fiction/algorithms/physical_design/post_layout_optimization.hpp

```diff
@@ -918,65 +364,92 @@ class post_layout_optimization_impl
                                   post_layout_optimization_stats& st) :
             plyt{lyt},
             ps{p},
-            pst{st}
+            pst{st},
+            start{std::chrono::high_resolution_clock::now()}
     {}
 
     void run()
     {
         static_assert(is_gate_level_layout_v<Lyt>, "Lyt is not a gate-level layout");
         static_assert(is_cartesian_layout_v<Lyt>, "Lyt is not a Cartesian layout");
 
+        // start the stopwatch to measure total optimization time
         const mockturtle::stopwatch stop{pst.time_total};
+
+        // record initial layout statistics
         pst.x_size_before        = plyt.x() + 1;
         pst.y_size_before        = plyt.y() + 1;
         pst.num_wires_before     = plyt.num_wires() - plyt.num_pis() - plyt.num_pos();
         pst.num_crossings_before = plyt.num_crossings();
 
-        uint64_t max_gate_relocations = ps.max_gate_relocations.value_or((plyt.x() + 1) * (plyt.y() + 1));
+        // determine the maximum number of gate relocations
+        max_gate_relocations = ps.max_gate_relocations.value_or((plyt.x() + 1) * (plyt.y() + 1));
+
+        // create an obstruction layout based on the original layout
+        auto layout = obstruction_layout<Lyt>(plyt);
 
-        // optimization
-        auto layout                  = obstruction_layout<Lyt>(plyt);
+        // initialize flags to control the optimization loop
         bool moved_at_least_one_gate = true;
         bool reduced_wiring          = true;
+        timeout_limit_reached        = false;
 
         if (max_gate_relocations == 0)
         {
-            fiction::wiring_reduction_stats wiring_reduction_stats{};
-            fiction::wiring_reduction(layout, &wiring_reduction_stats);
+            // perform wiring reduction only when no gate relocations are allowed
+            fiction::wiring_reduction_stats  wiring_reduction_stats{};
+            fiction::wiring_reduction_params wiring_reduction_params{};
+
+            // update the remaining timeout
+            wiring_reduction_params.timeout = update_timeout();
+
+            if (!timeout_limit_reached)
+            {
+                fiction::wiring_reduction(layout, wiring_reduction_params, &wiring_reduction_stats);
+            }
         }
         else
         {
-            while (moved_at_least_one_gate || reduced_wiring)
+            // iteratively optimize the layout until no more improvements can be made or timeout is reached
+            while ((moved_at_least_one_gate || reduced_wiring) && !timeout_limit_reached)
             {
                 reduced_wiring = false;
-                fiction::wiring_reduction_stats wiring_reduction_stats{};
-                if (moved_at_least_one_gate && !ps.optimize_pos_only)
+
+                // update the remaining timeout
+                fiction::wiring_reduction_params wiring_reduction_params{};
```

Comment:
See above.

Also, why the duplication? Can we make these params a member of the class?

Reply:
Same with the statistics then.